1445 sql primary keys (#1465)
Codecademy/docs: content/sql/concepts/primary-keys/primary-keys.md

  - 'paths/analyze-data-with-sql'
---

**Primary keys** are special columns that are used to uniquely identify each row of a table in SQL.

A primary key column has a few requirements:

- None of the values can be `NULL`.
- Each value must be unique (e.g., two rows in a `customers` table wouldn't have the same primary `customer_id`).

> - A table can not have more than one primary key column.

Reviewer: Is this actually the case given the new information about composite keys?

Author: Hi Dusch! I believe the statement is accurate given the example we're studying there. Mentioning composite keys or how a field or column could appear with the same value in different rows at this moment is—in my opinion—something that would confuse the reader. Besides the concept itself, composite keys are known for being not the best design, performance-wise. I think the doc is super clear as it is now!

Reviewer: Gotcha! Any thoughts, @SSwiniarski?

Reviewer 2: @Dusch4593 I think we should delete the word "column" so it just says "A table cannot have more than one primary key."

Reviewer: ```suggestion
- A table can not have more than one primary key.
```

  - 'paths/analyze-data-with-sql'
---

> Sometimes, SQL tables are distinguished through **primary keys**, a special column that will uniquely identify each row of that table.

Reviewer: ```suggestion
**Primary keys** are special columns are used to uniquely identify each row of a table in SQL.
```

Author: Hey Dusch! I Guess you meant: "are special columns **that** are used". I will modify and submit accordingly! Thanks!

When the primary key for one table appears in a different table, it is called a foreign key.

Why is this important? The most common types of joins will be joining a foreign key from one table with the primary key from another table. For instance, when we join the `orders` table and the `customers` table, we join on the `customer_id` column, which is a foreign key in `orders` and the primary key in `customers`.

## Composite Keys

> Having one primary key per table is sometimes not enough to uniquely identify a row; in such cases we would use more than one column, or **composite keys**. This requirement should be detected during the designing phase of a database. 

Reviewer: ```suggestion
Sometimes, having one primary key per table is not enough to uniquely identify a row. In such cases, multiple columns would work as composite keys for the table. This requirement should be detected during the designing phase of a database.
```

Reviewer: No need to **emphasize** here; only the concept/term matching the `Title` should be bolded once, near the top of the file.

> For example, a database containing vehicle parts will have to uniquely identify a row of parts. You could use either the `engine_ID` or  `body_ID`; however, this may create ambiguity as cars could get their engines swapped.
> 
> Usually (but depending on local regulations) a car will require an engine part ID and a car body ID to be associated with a license plate. In this case you could have a table with more information about the car, such as `left_door_ID`, `gearbox_ID`, etc., but you would identify a specific car by two different aspects: its body and its engine. 
> 
> A composite key would be useful in this case. This is how a `vehicle_registry` table would look (omitted extra parts/columns for brevity):

Reviewer: ```suggestion
For example, a database of car parts will have to uniquely identify a row of parts. You could use either the `engine_ID` or  `body_ID`; however, this may create ambiguity as cars could get their engines swapped.

Depending on local regulations, a car may require an engine part ID and a body ID to be associated with a license plate. One solution might be adding more row information about the car, such as `left_door_ID`, `gearbox_ID`, etc. But then a specific car would have to be identified by two different aspects: its body and its engine. 

A composite key would be useful in this case. This is how a `vehicle_registry` table might look (extra parts/columns omitted for brevity):
```
